lib/chef/util/dsc/resource_store.rb

```diff
@@ -0,0 +1,114 @@
+#
+# Author:: Jay Mundrawala (<jdm@chef.io>)
+#
+# Copyright:: Copyright (c) 2015 Chef Software, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef/util/powershell/cmdlet'
+require 'chef/util/powershell/cmdlet_result'
+
+class Chef
+class Util
+class DSC
+  class ResourceStore
+
+    def self.instance
+      @@instance ||= ResourceStore.new.tap do |store|
+        store.send(:populate_cache)
+      end
+    end
+
+    def resources
+      @resources ||= []
+    end
+
+    def resource(name, module_name=nil)
+      found = find_resources(name, module_name, resources)
+
+      # We don't have it, query for the resource...it might
+      # have been added since we last queried
+      if found.length == 0
+        rs = query_resource(name)
+        add_resources(rs)
+        found = find_resources(name, module_name, rs)
+      end
+
+      case found.length
+      when 0
+        nil
+      when 1
+        found[0]
+      else
+        raise 'Multiple matching resources'
```

Comment:
Presumably we never get here, but if we did it might help to know what we found. Could we include the name we matched in here? like "DSC resource '#{name}' matched multiple resources unexpectedly"?